pkg/pillar/cas/containerd.go

```diff
@@ -159,34 +160,55 @@ func (c *containerdCAS) IngestBlob(blobs ...*types.BlobStatus) ([]*types.BlobSta
 		}
 		log.Infof("IngestBlob(%s): Attempting to load blob", blob.Sha256)
 		var (
-			r        io.Reader
-			blobFile = blob.Path
+			r, contentReader io.Reader
+			err              error
+			closer           io.Closer
+			blobFile         = blob.Path
 			// the sha MUST be lower-case for it to work with the ocispec utils
 			sha = fmt.Sprintf("%s:%s", digest.SHA256, strings.ToLower(blob.Sha256))
 		)
 
-		//Step 1.1: Read the blob from verified dir
-		fileReader, err := os.Open(blobFile)
-		if err != nil {
-			err = fmt.Errorf("IngestBlob(%s): could not open blob file for reading at %s: %+s",
+		//Step 1.1: Read the blob from verified dir or provided content
+		switch {
+		case blobFile == "" && len(blob.Content) == 0:
+			err = fmt.Errorf("IngestBlob(%s): both blobFile and blobContent empty %s: %+s",
+				blob.Sha256, blobFile, err.Error())
+			log.Errorf(err.Error())
+			return loadedBlobs, err
+		case blobFile != "" && len(blob.Content) != 0:
+			err = fmt.Errorf("IngestBlob(%s): both blobFile and blobContent provided, cannot pick, %s: %+s",
 				blob.Sha256, blobFile, err.Error())
 			log.Errorf(err.Error())
 			return loadedBlobs, err
+		case blobFile != "":
+			fileReader, err := os.Open(blobFile)
+			if err != nil {
+				err = fmt.Errorf("IngestBlob(%s): could not open blob file for reading at %s: %+s",
+					blob.Sha256, blobFile, err.Error())
+				log.Errorf(err.Error())
+				return loadedBlobs, err
+			}
+			defer fileReader.Close()
+			contentReader = fileReader
+			closer = fileReader
+		default:
+			contentReader = bytes.NewReader(blob.Content)
 		}
-		defer fileReader.Close()
 
 		//Step 1.2: Resolve blob type and if this is a manifest or index, we will need to process (parse) it accordingly
 		switch {
 		case blob.IsIndex():
 			// read it in so we can process it
-			data, err := ioutil.ReadAll(fileReader)
+			data, err := ioutil.ReadAll(contentReader)
 			if err != nil {
 				err = fmt.Errorf("IngestBlob(%s): could not read data at %s: %+s",
 					blob.Sha256, blobFile, err.Error())
 				log.Errorf(err.Error())
 				return loadedBlobs, err
 			}
-			fileReader.Close()
+			if closer != nil {
+				closer.Close()
+			}
```

Comment:
Should this be in a defer statement up front?

Reply:
Yeah, I like that better. I will do that.